### DISCUSSION TITLE
QuadraturePointMultiapp not working among parallel cores

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#20749] [Q&A General] [answered]
Hi,
The new multi-app system that i am developing based on the element quadrature points does not seem to work when executed in parallel. The issue mainly stems from the RVE sub-model where the simulation fails to converge. The problem could potentially lie in the object `_positions` which collects the coordinates of the individual quadrature points. The following code block is what i am having in the class QuadraturePointMultiapp.
```
void
QuadraturePointMultiapp::fillPositions()
{
  MooseMesh & master_mesh = _fe_problem.mesh();
  auto & mesh = master_mesh.getMesh();
  for (auto & elem : mesh.active_local_element_ptr_range()){
    const FEFamily mapping_family = FEMap::map_fe_type(*elem);
    
    // Get a constant reference to the Finite Element type
    FEType fe_type(elem->default_order(),mapping_family);

    // Build a Finite Element object of the specified type.  Since the
    // FEBase::build() member dynamically creates memory we will
    // store the object as a std::unique_ptr<FEBase>.  This can be thought
    // of as a pointer that will clean up after itself.
    std::unique_ptr<FEBase> fe(FEBase::build(elem->dim(), fe_type));
    const int extraorder = 0;
    std::unique_ptr<QBase> qrule (fe_type.default_quadrature_rule (elem->dim(), extraorder));
  
    // Tell the finite element object to use our quadrature rule.
    fe->attach_quadrature_rule (qrule.get());

    // The physical XY locations of the quadrature points on the element.
    // These might be useful for evaluating spatially varying material
    // properties at the quadrature points.
    const std::vector<Point> & q_points = fe->get_xyz();
    fe->reinit(elem);
      
    if (hasBlocks(elem->subdomain_id())){
       for (const auto & q: q_points)
         _positions.push_back(q);
    }
  }

  // Use the comm from the problem this MultiApp is part of
  libMesh::ParallelObject::comm().allgather(_positions);

  if (_positions.empty())
    mooseError("No positions found for QuadraturePointMultiapp ", _name);

  // An attempt to try to make this parallel stable
  std::sort(_positions.begin(), _positions.end());
}
```
I do not see any issues here as the code simply collects the QP coordinates and store that on to this array `_positions`. If i ask the code to use the centroidal coordinates instead of the actual QP location it works fine. But somehow the parallel computation does not like this array which leads to the non-convergence of sub-app. I wonder if there is anything wrong with the way the QP coordinates are calculated, which in my opinion is very straightforward. Would be really appreciated if someone could point me to the right direction.
Kind regards,
Arun

**Comment by GiudGiud** [comment, marked as answer]
Hello

I also dont see what's wrong with this. 
Could please attach more information to this post?
Let's have a log with 2 MPI processes, a few (maybe like 10) quadrature points at most. And let's see the subapp convergence issues.

Does everything work well in serial?
What transfers do you have to the subapps right now? Do they run on their own or do you need the transfer to happen to get convergence?

Guillaume

**Reply by abarun22** [reply]
Hi,
Here is an example simulation with one element and 2 mpi processes. As you can see from the attached log, all goes well until the second macro time step, after which the sub-app cease to converge. The serial run is absolutely fine. At present i do not intend to transfer any thing out of the main model. Wanted to make sure the QuadraturePointMultiapp works OK in parallel and then think about including them.
Here are the contents of my main model if you would like to look at.
```
# Model at macro scale used to analyse larger components
# Uniaxial tension with finite strain elasticity

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Mesh]
    type = GeneratedMesh
    dim = 3
    nx = 1
    ny = 1
    nz = 1
[]

[Modules/TensorMechanics/Master]
  [./block1]
    strain = FINITE
    add_variables = true
#    use_automatic_differentiation = true
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 2.1e5
    poissons_ratio = 0.3
  [../]
  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]
[]

[BCs]
  [./left_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'left'
    value = 0.0
  [../]
  [./left_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'left'
    value = 0.0
  [../]
  [./left_z]
    type = DirichletBC
    variable = disp_z
    boundary = 'left'
    value = 0.0
  [../]

  [./right]
    type = FunctionDirichletBC
    variable = disp_x
    boundary = 'right'
    function = '0.01*t'
  [../]
[]

[AuxVariables]
  [./stress_xx_av_qp1]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[AuxKernels]
  [./stress_xx_qp1]
    type = RankTwoAux
    rank_two_tensor = stress
    variable = stress_xx_av_qp1
    index_i = 0
    index_j = 0
    selected_qp = 0
  [../]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  num_steps = 5
  dt = 0.01
  solve_type = 'PJFNK'
  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-pc_type -sub_pc_type -pc_asm_overlap -ksp_gmres_restart'
  petsc_options_value = 'asm lu 1 101'
[]

[Outputs]
   file_base = main
   type = Exodus
[]

[MultiApps]
  [sub]
    type = QuadraturePointMultiapp
    input_files = 'sub_gs_uniaxial_noCPFE.i'
    execute_on = timestep_end
    reset_time = '0.02 0.03 0.04 0.05'
  []
[]

```
Kind regards,
Arun

**Reply by GiudGiud** [reply]
Hello

Can we see the log for the divergence of the subapps?
Do they converge when run outside of the multiapp setup?
Also please keep n_mpi > n_elements. MOOSE isnt built for these edge cases. 

Guillaume

**Reply by abarun22** [reply]
Hi Guillaume,
Sorry about that. Here is the log for the divergence of sub-model. It does converge OK when executed in parallel and independently of the multi-app set-up. That's fine keeping n_mpi > n_elements, which was the case originally.
[run.log](https://github.com/idaholab/moose/files/8495925/run.log)

**Reply by GiudGiud** [reply]
so this isnt a real divergence. A residual of 1e-12 is good enough for most physics. It's just struggling to get any lower because the system is ill-conditioned, or because of other sources of numerical precision losses

You can see if you get lower by setting `automatic_scaling=true` in [Executioner]
but otherwise I would just set
`nl_abs_tol = 1e-11` in that same block

**Reply by abarun22** [reply]
Yes, its just a matter of setting the right tolerance. With `automatic_scaling=true` the sub-app runs fine and converges. My another query is on the selection of number of apps to execute this problem. Being a single element 3D model, i expect the multiapp system to create 8 sub-apps, each one dedicated for one quadrature point, which is OK for a serial run. The parallel execution shows only creation of 4 sub-apps which is quite strange. Any thoughts on this?

**Reply by GiudGiud** [reply]
the logging might be simply hiding them. We only output to console for rank 0
Add --keep-cout to the command line arguments if you want logs from all MPI processes

**Comment by abarun22** [comment]
With this CL option i could able to see the log for all processes, but only first 4 of them producing results, while other's showing as converged but does not actually undergoes any solution iterations. I believe these might've run correctly, but the outputting may not always be a true representation of actual. Attached the log for your reference.
[run.log](https://github.com/idaholab/moose/files/8497142/run.log)

**Reply by GiudGiud** [reply]
I think we can trust the output here. 
The subapps on rank 2 start with 0 residual. Any reason for that? What makes the apps have a non zero residual on rank 1?

**Comment by abarun22** [comment]
What timestep are you referring to? The first two timesteps seems ok as we observe non-zero residuals for all the sub-apps. For the subsequent timesteps sub-apps with rank 0-3 shows a non-zero residual while for the rest zero residuals were calculated. Its hard to interpret these print outputs as it shows differently for each selection of number of cores. When executed with 8 cores you get a different mode of output as attached. 
[run.log](https://github.com/idaholab/moose/files/8505862/run.log)
The right way to check the validity of the results would be to look at the exodus output and see if they make sense. I am not pretty sure you agree with me here.

**Reply by GiudGiud** [reply]
yes that's a good idea and that should work. 
If you are not transferring any quantity, I would expect all 8 subapps to behave exactly the same

**Reply by abarun22** [reply]
Yes that's right

**Reply by GiudGiud** [reply]
@abarun22 could you confirm the results make sense here?
- All 8 subapps give the same results without transfers
- The time dependence of the solves without transfers are as expected (eg converges to something then doesnt move from timestep to timestep)

**Reply by GiudGiud** [reply]
@abarun22 any update on this?

**Reply by GiudGiud** [reply]
btw I'd love the QuadratureMultiApp to make it into the framework if it's working satisfactorily

**Reply by abarun22** [reply]
Hi Guillaume,
Yes, absolutely. Without the transfers all the sub-apps produce the same set of results and these are observed to be consistent between the time steps. The attached execution log will tell you the story. 
Let me know if you need further clarification.
Kind regards,
Arun
[run.log](https://github.com/idaholab/moose/files/8877734/run.log)

**Reply by GiudGiud** [reply]
ok so what is the current issue?

Actually looking at the log I m a little confused. Why are the outputs for the app initializations staggered with the solves?
All the initializations should happen at the beggining then all the solves.

Guillaume

**Reply by abarun22** [reply]
Well i do not see any problems here. The output sub-app initializations seems to be in order i.e., the initializations followed by the solves.  A staggered outpput is possible only if you forcibly try to output the log for all processors. Currently it only outputs the log for zeroth rank and rest of them are suppressed and so you get an ordered output. Are there any other concerns with this work before it gets merged with the framework code?

**Reply by GiudGiud** [reply]
Ok all good. I was asking you if you were having issues.

If you have time to make a pull request with the QuadraturePointMultiapp, some docs and testing then we can review it and consider merging it in the framework

**Reply by abarun22** [reply]
I think this one is a pretty straight forward work to be added in to the framework. All it does is the creation of sub-apps at the main app quadrature point locations and execute them. The only issue i was facing earlier is the non-convergence of sub models for a parallel execution which was corrected with the `automatic_scaling` option. Otherwise it should be good to go. I shall create a PR very soon and let you know.
